### ISSUE TITLE
JSON document inspection API needs some love

### ISSUE THREAD

**Issue by sbseltzer** [inmotionsoftware/ims-json#2]
The [README](README.md) includes no examples of inspecting a parsed document. I've had to look through the source code to figure it out, but it still feels clunky and somewhat obscured by the macros. The document building C API is gorgeous and perhaps my favorite of anything I've seen so far (and I've evaluated dozens of JSON libraries). It'd be best if the document inspection half of the API felt symmetrical to the document building half. Right now that does not seem to be the case.

So this issue comes in two parts.
1. Document the current inspection strategy.
2. Come up with an improved one that mimics the way documents are built.

Let me know if you'd like help with these. I'll spend some time thinking about what the inspection API might look like, but I'm guessing you already have something in mind.

**Comment by bghoward** [comment]
Thanks for your feedback. If you have some ideas on what you'd want to see, It'd be helpful. I can spend some time documenting the inspection side of things.

**Comment by sbseltzer** [comment]
Indeed, only the C++ inspection is documented. Thanks for being so responsive!

I'll do some thinking out loud here so we can brainstorm this.

There's something about the `find` vs `get` semantics that feel a little confusing. I'm also not sure how I feel about the `jobj_find` naming convention on the macros. It's not consistent with `jarray_get`. It sorta makes sense being that they are looking for a value that may not exist, but I'm used to seeing `get` a lot. That's just a nitpick, though.

Another oddity is why the `*_get_str` macros have been commented out.

The `jobj_find_*` all have default return values for when the value of the specified type isn't found, but those could be misleading.
```
bool yes = jobj_find_bool(obj, "yes");
```
If this returns false, did it mean `"yes"` wasn't there, `"yes"` was another type, or `"yes"` was set to false? Knowing that requires the user to do 3 more things.

1. Get the `jval_t` version using `jobj_find` so that it can be inspected.
2. Check if it existed by comparing to `JNULL_VAL`.
3. Check whether it was a bool with `jval_is_bool`.

If the user wants to program defensively, the `jobj_find_*` macros are significantly less useful, and are better off doing the above followed by `json_get_bool` (somewhat unintuitive as I would've thought it'd be `jval_get_bool` or `jval_as_bool`, but I can see why you did this).

To summarize, handling this:
```
{
  bool yes = jobj_find_bool(obj, "yes");
}

```
Expands to this for the paranoid user:
```
{
  bool yes = false;
  jval_t yesval = jobj_find(obj, "yes");
  if (yesval == JNULL_VAL)
  {
    /* handle missing keyvalue */
  }
  else if (!jval_is_bool(yesval))
  {
    /* handle type mismatch */
  }
  yes = json_get_bool(jsn, yesval);
}
```
But could perhaps look more like this:
```
{
  jval_t val;
  bool yes = jobj_get_bool(obj, "yes", &val);
  if (val == JNULL_VAL)
  {
    /* handle missing keyvalue */
  }
  else if (!jval_is_bool(val))
  {
    /* handle type mismatch */
  }
}
```
Although the last one is merely saving one extra line, it gives you everything you need right out the gate to inspect if you so choose. It's also a little more control-flow friendly IMO - if the user really doesn't care, pass NULL for the `jval_t *` and it becomes functionally the same as `jobj_find_bool`. I think the array specification could work similarly.

One thing I liked from parson and a number of frameworks was `dotget` functionality.
```
{ "obj" : { "key" : "value" } }
```
Perhaps value could be retrieved with
```c
const char *str = jobj_dotget_str(".obj.key");
```
or maybe similar to before
```c
jval_t val;
const char *str = jobj_dotget_str(".obj.key", &val);
```
Let me know what you think of these ideas.
